File: apps/ruby-sdk/lib/firecrawl/models/product_profile.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Structured product information extracted from a product page via the
    # `product` scrape format.
    class ProductProfile
      # An image associated with a product or variant.
      class Image
        attr_reader :url, :alt

        def initialize(data)
          @url = data["url"]
          @alt = data["alt"]
        end
      end

      # A monetary value with an optional currency and formatted string.
      class Price
        attr_reader :amount, :currency, :formatted

        def initialize(data)
          @amount = data["amount"]
          @currency = data["currency"]
          @formatted = data["formatted"]
        end
      end

      # Stock availability information for a variant. Always present.
      class Availability
        attr_reader :in_stock, :text

        def initialize(data)
          data ||= {}
          @in_stock = data["inStock"] || false
          @text = data["text"]
        end
      end

      # Sale pricing for a variant, carrying the pre-sale original price.
      class Sale
        attr_reader :original_price

        def initialize(data)
          @original_price = data["originalPrice"] && Price.new(data["originalPrice"])
        end
      end

      # A purchasable variant of a product. Pricing, availability, and images
      # live here rather than on the top-level product.
      class Variant
        attr_reader :id, :sku, :title, :values, :price, :sale,
                    :availability, :images

        def initialize(data)
          @id = data["id"]
          @sku = data["sku"]
          @title = data["title"]
          @values = data["values"]
          @price = data["price"] && Price.new(data["price"])
          @sale = data["sale"] && Sale.new(data["sale"])
          @availability = Availability.new(data["availability"])
          @images = (data["images"] || []).map { |img| Image.new(img) }
        end
      end

      attr_reader :title, :brand, :category, :url, :description, :variants

      def initialize(data)
        @title = data["title"]
        @brand = data["brand"]
        @category = data["category"]
        @url = data["url"]
        @description = data["description"]
        @variants = (data["variants"] || []).map { |variant| Variant.new(variant) }
      end

      def to_s
        "ProductProfile{title=#{title || 'untitled'}, url=#{url || 'unknown'}}"
      end
    end
  end
end

File: apps/dot-net-sdk/Firecrawl/Models/ProductProfile.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Firecrawl.Models;

/// <summary>
/// Structured product information extracted via the <c>product</c> scrape format.
/// </summary>
public class ProductProfile
{
    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("brand")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Brand { get; set; }

    [JsonPropertyName("category")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Category { get; set; }

    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("description")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Description { get; set; }

    [JsonPropertyName("variants")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<ProductVariant>? Variants { get; set; }
}

/// <summary>
/// An image associated with a product or product variant.
/// </summary>
public class ProductImage
{
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("alt")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Alt { get; set; }
}

/// <summary>
/// A monetary price for a product or product variant.
/// </summary>
public class ProductPrice
{
    [JsonPropertyName("amount")]
    public double Amount { get; set; }

    [JsonPropertyName("currency")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Currency { get; set; }

    [JsonPropertyName("formatted")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Formatted { get; set; }
}

/// <summary>
/// Availability information for a product or product variant.
/// </summary>
public class ProductAvailability
{
    [JsonPropertyName("inStock")]
    public bool InStock { get; set; }

    [JsonPropertyName("text")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Text { get; set; }
}

/// <summary>
/// A purchasable variant of a product (e.g. a specific size/color combination).
/// </summary>
public class ProductVariant
{
    [JsonPropertyName("id")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Id { get; set; }

    [JsonPropertyName("sku")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Sku { get; set; }

    [JsonPropertyName("title")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Title { get; set; }

    [JsonPropertyName("values")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Dictionary<string, JsonElement>? Values { get; set; }

    [JsonPropertyName("price")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ProductPrice? Price { get; set; }

    [JsonPropertyName("sale")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ProductSale? Sale { get; set; }

    [JsonPropertyName("availability")]
    public ProductAvailability Availability { get; set; } = new();

    [JsonPropertyName("images")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<ProductImage>? Images { get; set; }
}

/// <summary>
/// Sale information for a product variant (the price before the current sale).
/// </summary>
public class ProductSale
{
    [JsonPropertyName("originalPrice")]
    public ProductPrice OriginalPrice { get; set; } = new();
}
